pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/table/DynamicPulsarSerializationSchema.java

```diff
@@ -192,10 +196,25 @@ public void setNumParallelInstances(int numParallelInstances) {
         if (StringUtils.isBlank(valueFormatType)) {
             return new FlinkSchema<>(Schema.BYTES.getSchemaInfo(), valueSerialization, null);
         }
-        SchemaInfo schemaInfo = SchemaUtils.tableSchemaToSchemaInfo(valueFormatType, valueDataType);
+        Map<String, String> options = new HashMap<>();
+        hackPbSerializationSchema(options);
+        SchemaInfo schemaInfo = SchemaUtils.tableSchemaToSchemaInfo(valueFormatType, valueDataType, options);
         return new FlinkSchema<>(schemaInfo, valueSerialization, null);
     }
 
+    private void hackPbSerializationSchema(Map<String, String> options) {
+        // reflect read PbRowSerializationSchema#messageClassName
+        if (valueSerialization instanceof PbRowDataSerializationSchema) {
+            try {
+                String messageClassName =
+                        (String) FieldUtils.readDeclaredField(valueSerialization, "messageClassName", true);
```

Comment:
"messageClassName" declare  as  static final class-field ?

Reply:
Well well, it can be declared as final.

Reply:
I added the schema cache so that this method will only be called once